src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/StringOperatorsTest.java

```diff
@@ -0,0 +1,197 @@
+/*
+ *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.unittest;
+
+import com.amazon.opendistroforelasticsearch.sql.parser.ScriptFilter;
+import com.amazon.opendistroforelasticsearch.sql.parser.SqlParser;
+import com.amazon.opendistroforelasticsearch.sql.util.CheckScriptContents;
+import org.elasticsearch.search.builder.SearchSourceBuilder.ScriptField;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import static org.junit.Assert.assertTrue;
+
+public class StringOperatorsTest {
+
+    private static SqlParser parser;
+
+    @BeforeClass
+    public static void init() { parser = new SqlParser(); }
+
+    @Test
+    public void substringTest() {
+        String query = "SELECT substring(lastname, 2, 1) FROM accounts WHERE substring(lastname, 2, 1) = 'a' " +
+                "GROUP BY substring(lastname, 2, 1) ORDER BY substring(lastname, 2, 1)";
+
+        ScriptField scriptField = CheckScriptContents.getScriptFieldFromQuery(query);
+        assertTrue(
+                CheckScriptContents.scriptContainsString(
+                        scriptField,
+                        "doc['lastname'].value.substring(1, 2)"));
+
+        ScriptFilter scriptFilter = CheckScriptContents.getScriptFilterFromQuery(query, parser);
+        assertTrue(
+                CheckScriptContents.scriptContainsString(
+                        scriptFilter,
+                        "doc['lastname'].value.substring(1, 2)"
+                )
+        );
+    }
+
+    @Test
+    public void substringIndexOutOfBoundTest() {
+        String query = "SELECT substring('sampleName', 0, 20) FROM accounts";
+        ScriptField scriptField = CheckScriptContents.getScriptFieldFromQuery(query);
+        assertTrue(
+                CheckScriptContents.scriptContainsString(
+                        scriptField,
+                        "'sampleName'.substring(0, 10)"
```

Comment:
Should the second argument be 20? Did this test pass?

Reply:
@dai-chen the es x-pack will not throw exception when the index are out of bound execute it like the endIndex is corrected to the index of the last position in a string, but the painless language does behave in that way. I correct the out-of-bound indexes to in-bound so that it behaves like x-pack.

Reply:
Got it. Thanks!